influxql/ast.go

```diff
@@ -1235,6 +1235,71 @@ func (s *SelectStatement) RewriteFields(ic IteratorCreator) (*SelectStatement, e
 	return other, nil
 }
 
+var (
+	// Exactly matches ^foo$ or ^$
+	exactMatchRegex = regexp.MustCompile(`^\^.*\$$`)
+)
+
+// RewriteRegexExprs rewrites regex conditions to make beter use of the database
+// index.
+//
+// Conditions that can currently be simplified are:
+//
+//     - host ~= /^foo$/ becomes host = 'foo'
+//     - host ~! /^foo$/ becomes host != 'foo'
+//
+// Note: If a backslash (\) is detected in the condition's value, then the
+// regex will *not* be rewritten. In order to support rewriting regexes with
+// escaped characters we need to write a parser, rather than try to match with
+// a regex.
+func (s *SelectStatement) RewriteRegexConditions() {
+	s.Condition = RewriteExpr(s.Condition, func(e Expr) Expr {
+		be, ok := e.(*BinaryExpr)
+		if !ok || (be.Op != EQREGEX && be.Op != NEQREGEX) {
+			// This expression is not a binary condition or doesn't have a
+			// regex based operator.
+			return e
+		}
+
+		// Handle regex-based condition.
+		var matched bool
```

Comment:
The reason it's there is in preparation for when I add further regex rewrites along side this exact match rewrite. Once they're added the `defer` block will be cleaner since there will be multiple matched return points. Happy to remove it until then though, if it's complicating things 👍